### ISSUE TITLE
Unable to prevent reportUndefinedVariable false positives

### ISSUE THREAD

**Issue by hmeine** [microsoft/pylance-release#1203]
## Environment data

-   Language Server version: 2021.4.2 (pyright 3888d715)
-   OS and version: macOS 11.3
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.9 from MacPorts

## Expected behaviour

I am getting `"ctx" is not defined Pylance(reportUndefinedVariable)` errors from code files where `ctx` is a symbol that's injected by the runtime environment (MeVisLab) for which these files are written.

I would have expected a way to suppress these false positives, much like with other tools: For instance, I have `additional-builtins` in my `.pylintrc`, and I have an environment variable `PYFLAKES_BUILTINS` set up.

Furthermore, I would have expected this configuration to be discoverable. It may also be a good idea to document that configuration for other tools like pylint or pyflakes does *not* apply, but that pylance is a standalone tool that does not make use of these. (Since some python linting tools call others, it was not obvious enough to me that pylance is not just some efficient integration of existing tools for VS Code.)

## Actual behaviour

* The measures I took to make pylint and pyflakes happy do not seem to apply to pylance, but I get a lot of false positives (with apparent side-effects for variable renaming, for instance?).
* clicking on the `reportUndefinedVariable` link does not lead to instructions how to configure additional builtins.
* I cannot find a way to configure these by searching for "pylance" in the VS Code configuration.
* `TROUBLESHOOTING.md` does not help, either.

## Logs

```
Background analysis message: setFileOpened
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (9ms)
[FG] parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 0ms] (103ms)
[FG] binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/builtins.pyi (115ms)
[FG] binding: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (1ms)
Background analysis message: getSemanticTokens full
[BG(1)] getSemanticTokens full at /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py ...
[BG(1)]   parsing: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (16ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 0ms] (130ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/builtins.pyi (60ms)
[BG(1)]   binding: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (4ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/typing.pyi [fs read 1ms] (104ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/typing.pyi (38ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/re.pyi [fs read 0ms] (8ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/re.pyi (11ms)
[BG(1)]   parsing: /Users/hmeine/Developer/minimal_mevis/mevis.py [fs read 0ms] (4ms)
[BG(1)]   binding: /Users/hmeine/Developer/minimal_mevis/mevis.py (1ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/abc.pyi [fs read 0ms] (1ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/abc.pyi (0ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/enum.pyi [fs read 1ms] (9ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/enum.pyi (1ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/sys.pyi [fs read 0ms] (31ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/sys.pyi (2ms)
[BG(1)]   parsing: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/types.pyi [fs read 0ms] (11ms)
[BG(1)]   binding: /Users/hmeine/.vscode/extensions/ms-python.vscode-pylance-2021.4.2/dist/typeshed-fallback/stdlib/types.pyi (2ms)
[BG(1)] getSemanticTokens full at /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (750ms)
Background analysis message: getSemanticTokens range
[BG(1)] getSemanticTokens range 0:0 - 116:68 at /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (11ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py ...
[BG(1)]   checking: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (7ms)
[BG(1)] analyzing: /Users/hmeine/Developer/MeVisLab/Standard/Modules/Macros/Helpers/RunPythonScript.py (7ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

**Comment by jakebailey** [comment]
Pylance doesn't support adding random variables to the builtins (and I didn't know pylint/pyflakes have configuration for this).

Can you do something like this to notify type checkers (like pylance) that the additional variables exist?


```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    ctx: SomeType = ...
```

**Comment by hmeine** [comment]
That's a nice idea that would theoretically work, but we're talking about literally thousands of files, for which this boilerplate could would not be acceptable (most developers are not even using pylance).

The possibility to configure custom type stub files cannot be (mis)used for this, can it?  (I noticed in the above trace log that `typeshed-fallback/stdlib/builtins.pyi` contains builtins, but of course their existence was already known to pylance beforehand.)

**Comment by erictraut** [comment]
Libraries that "inject" symbols into the namespace of other modules are going to be problematic for any static type checking tool. (The practice is problematic for many reasons, including namespace pollution. I strongly recommend that libraries avoid this technique.)

Here are some options for you to consider:
1. Modify your code to explicitly import these symbols. Even if the symbols are being implicitly injected into your modules' namespace at runtime, making the import explicit will have other benefits.
2. Globally disable the "reportUndefinedVariable" diagnostic rule for your project to silence the error.
3. You could clone the typeshed stdlib stubs (e.g. copy them out of the pylance bundle or check out a fresh copy from the typeshed repo) and modify the builtins.pyi file to include your additional symbols. This is pretty fragile though, and you would need to occasionally sync to the latest contents of typeshed to stay current. I personally wouldn't recommend this approach, but I figured I'd list it for completeness.

**Comment by hmeine** [comment]
Yes, I also agree that this injection is not nice. On the other hand, MeVisLab is 20 years old and there are too many files relying on this behavior. (It's also only exactly this one symbol `ctx` for the local *context*, so it's not much in terms of pollution. Other functionality is just imported.)

TL;DR: 1 is no option here, even if I would prefer it otherwise.

Since I like the "reportUndefinedVariable" functionality a lot, I'll have to investigate option 3, even though it is a hack.

**Comment by savannahostrowski** [comment]
If you're investigating option 3, I'm going to close this issue as I don't think that this is waiting for user response anymore. Please ping if you think this should be reopened.

**Comment by hmeine** [comment]
Yes, makes sense.

For the record, I wrote a script to automate locating and modifying `builtins.pyi` which I run after each VS Code upgrade.
